### PR TITLE
MRG: Faster topomaps

### PR DESCRIPTION
**Pull request by Eric89GXL**
This is an effort to make our topomap plotting faster interactively and in testing. cc @jona-sassenhagen can you push your changes to the tests?

Still WIP because I need to modify the picking to use `PatchCollection` (which I *think* [should be possible](https://matplotlib.org/users/event_handling.html#object-picking)) instead of iteration over `Circle` objects. @jaeilepp did you ever try this approach?

### CODE REVIEW

**Comment by jona-sassenhagen** [comment]
>@jona-sassenhagen can you push your changes to the tests?

I'm still profiling, sorry ...

**Comment by Eric89GXL** [comment]
Okay, well if you get anything useful let me know. In the meantime feel free to try this branch, too.

**Comment by Eric89GXL** [comment]
Looks like it successfully cut down testing time by about 5 minutes. The longest `topomap` test is now only 15 sec:

https://travis-ci.org/mne-tools/mne-python/jobs/225709024#L3282

**Comment by jona-sassenhagen** [comment]
Awesome! With a bit of luck I'll be able to shave off another second or two :D

**Comment by Eric89GXL** [comment]
@jaeilepp I'm not going to work on this until tomorrow sometime, so if you feel like messing with the PatchCollection selection stuff in the meantime, feel free.

File: mne/time_frequency/tfr.py [hunk 1]
```diff
@@ -1127,11 +1127,9 @@ def plot(self, picks, baseline=None, mode='mean', tmin=None, tmax=None,
 
     def _onselect(self, eclick, erelease, baseline, mode, layout):
         """Handle rubber band selector in channel tfr."""
-        import matplotlib.pyplot as plt
         from ..viz import plot_tfr_topomap
         if abs(eclick.x - erelease.x) < .1 or abs(eclick.y - erelease.y) < .1:
             return
-        plt.ion()  # turn interactive mode on
```
Comment (Eric89GXL) [inline mne/time_frequency/tfr.py]: I don't think we should mess with matplotlib interactive status. Messing with global states is not a nice thing for a package to do.

It would be better to document any functions that need interactivity, telling people to activate interactive mode (maybe with docs somewhere about how to do it).

**Comment by Eric89GXL** [comment]
@kingjr feel free to try this to see if it works any better for you. If not, let me know which call(s) are slow for you.

File: mne/viz/topomap.py [hunk 2]
```diff
@@ -370,37 +370,30 @@ def _griddata(x, y, v, xi, yi):
...
 
 def _plot_sensors(pos_x, pos_y, sensors, ax):
     """Plot sensors."""
     from matplotlib.patches import Circle
+    from matplotlib.collections import PatchCollection
     if sensors is True:
+        patches = list()
         for x, y in zip(pos_x, pos_y):
-            ax.add_artist(Circle(xy=(x, y), radius=0.003, color='k'))
+            patches += [Circle(xy=(x, y), radius=0.003)]
```
Comment (jona-sassenhagen) [inline mne/viz/topomap.py]: Why not

```
patches = [Circle(xy=(x, y), radius=0.003) for x, y in zip(pos_x, pos_y)]
```

File: mne/viz/topomap.py [hunk 3]
```diff
@@ -370,37 +370,30 @@ def _griddata(x, y, v, xi, yi):
...
+        ax.add_collection(PatchCollection(patches, color='k'))
```
Comment (jona-sassenhagen) [inline mne/viz/topomap.py]: @jaeilepp it's probably suboptimal that the legend for spatial_colors and this one use different code ...

spatial_colors doing `ax.scatter(pos_x, pos_y, color=colors, s=25, marker='.', zorder=1)`

File: mne/viz/topomap.py (hunk 2 above)
Comment (Eric89GXL) [inline mne/viz/topomap.py]: Yeah that's better, just bad translation by me. Actually I'm not sure if this is really any better than `ax.plot()`, I should probably try that, too. (Need to check and see if the subset of selected points is also accessible there, though, just like with PatchCollection, though.)

File: mne/viz/topomap.py (hunk 3 above)
Comment (Eric89GXL) [inline mne/viz/topomap.py]: My guess is that it has to do with selection. These sensors need to be Lasso-able. The existing code makes it easy. But I think it should be possible at least with PatchCollection if not also whatever `scatter` returns with a bit of extra logic at the selection end.

File: mne/viz/topomap.py (hunk 2 above)
Comment (jona-sassenhagen) [inline mne/viz/topomap.py]: We should probably benchmark all 3(+?) methods we're using:
ax.plot
patches
ax.scatter

File: mne/viz/topomap.py (hunk 3 above)
Comment (Eric89GXL) [inline mne/viz/topomap.py]: ... although we'd have to check to see if you can do point-by-point color with `scatter` like you can with `PatchCollection`

Comment (jona-sassenhagen) [inline mne/viz/topomap.py]: I'm not sure I understand you correctly, but the (colored) legend for plot_joint is currently done with `scatter`.

Checking the history, your guess seems correct: I submitted this with add_artist patches and @jaeilepp turned it into scatter when he added interactivity.

Comment (Eric89GXL) [inline mne/viz/topomap.py]: I think you must have this backward, because right now it's `add_artist`, so you must have done `scatter` and @jaeilepp did `add_artist`...

But in any case, we should figure out which is fastest among `ax.plot`, `ax.scatter`, and `PatchCollection`. Then we can try from fastest to slowest to see if sub-selection can be done. Do you have time to do some profiling?

Comment (jona-sassenhagen) [inline mne/viz/topomap.py]: >I think you must have this backward, because right now it's add_artis

I meant this:
https://github.com/mne-tools/mne-python/blob/master/mne/viz/evoked.py#L167

>Do you have time to do some profiling?

Not today, I got to sleep :) I can set up a preliminary benchmark, but nothing thorough.

Comment (Eric89GXL) [inline mne/viz/topomap.py]: Well if you have time tomorrow morning feel free, otherwise I'll give it a shot

Comment (jona-sassenhagen) [inline mne/viz/topomap.py]: @Eric89GXL 
http://nbviewer.jupyter.org/gist/jona-sassenhagen/d46a30bc48bf5223bebeb827d2b20bb2

**Comment by Eric89GXL** [comment]
So `scatter`, `plot`, `PatchCollection`, and `master` ordered by speed. I don't think `plot` allows per-vertex colors but `scatter` and `PatchCollection` do. So now we just need to check if those (with `scatter` preferred) when picked give the indices of the subselection.

**Comment by jaeilepp** [comment]
> Still WIP because I need to modify the picking to use PatchCollection (which I think should be possible) instead of iteration over Circle objects. @jaeilepp did you ever try this approach?

No, I haven't really touched the topomap code too much.

**Comment by kingjr** [comment]
Seems like a x2 gain for me. Have you tried on the slider? There the gain should be noticeable.

**Comment by Eric89GXL** [comment]
Is that 2x enough to be happy, be acceptable, or be better but still crappy?

Is there an example or tutorial with the slider?

File: mne/viz/topomap.py (hunk 3 above)
Comment (jona-sassenhagen) [inline mne/viz/topomap.py]: I think the scatter option might give another decent boost.
Though most of the test comes from creating (and perhaps closing) the new axes, not the actual plot (it obscures the differences between the actual sensor plotting: I think scatter is actually ~2x as fast as the patches, not just a bit.)
So then it's possible we could save a lot with the axes divider pattern ..?

Comment (Eric89GXL) [inline mne/viz/topomap.py]: See also https://github.com/matplotlib/matplotlib/issues/6664. If it were faster to make axes I wouldn't have had to do crazy refactoring of the `topo` code to use a single Axes, either. But in the meantime, yeah using a single axes instance could help. But let's do it in another PR.

**Comment by kingjr** [comment]
> Is that 2x enough to be happy, be acceptable, or be better but still crappy?

yes, it's a good start

http://mne-tools.github.io/dev/auto_examples/visualization/make_report.html?highlight=slider

**Comment by Eric89GXL** [comment]
Okay now we use `SciPy` for the interpolation instead of using our own in-house radial basis function interpolation. I changed it not to use `Rbf` at all because it's slower than `CloughTocher2DInterpolator` but otherwise looks the same. In any case, things should be faster, especially in the case of animations and hopefully also in interactive mode (because we configure the interpolator once, more or less, instead of multiple times). Now the overhead is mostly in the plotting functions, and isn't too bad.

The example you listed @kingjr now rapidly pops up figures for me so I think the overhead is now mostly in figure/axis creation. Can you see if it got any faster for you?

Now I think I just need to make the selection bits work now that we use `scatter`, but it would be good if @jona-sassenhagen you could try it to make sure it's working well at your end, too.

**Comment by jona-sassenhagen** [comment]
@Eric89GXL I never use interactive functions, can't say anything to these.

In the notebook, I'm seeing a sizeable 30-50% speedup.

**Comment by Eric89GXL** [comment]
Okay, selection works now. Ready for review/merge from my end.

File: mne/viz/topomap.py [hunk 4]
```diff
@@ -363,44 +365,55 @@ def _draw_outlines(ax, outlines):
...
+class _GridData(object):
+    """Unstructured (x,y) data interpolator.
+
+    This class allows optimized interpolation by computing parameters
+    for a fixed set of true points, and allowing the values at those points
+    to be set independently.
+    """
+
+    def __init__(self, pos):
+        from scipy.spatial.qhull import Delaunay
```
Comment (jona-sassenhagen) [inline mne/viz/topomap.py]: Uhhhh Delaunay again, I recognize this now! This is like a second math ed to me!

File: mne/viz/topomap.py [hunk 5]
```diff
@@ -363,44 +365,55 @@ def _draw_outlines(ax, outlines):
...
 
 
 def _plot_sensors(pos_x, pos_y, sensors, ax):
     """Plot sensors."""
-    from matplotlib.patches import Circle
     if sensors is True:
-        for x, y in zip(pos_x, pos_y):
-            ax.add_artist(Circle(xy=(x, y), radius=0.003, color='k'))
+        ax.scatter(pos_x, pos_y, s=0.25, marker='o',
+                   edgecolor=['k'] * len(pos_x), facecolor='none')
```
Comment (jona-sassenhagen) [inline mne/viz/topomap.py]: `edgecolor='k'`?

Comment (Eric89GXL) [inline mne/viz/topomap.py]: I prefer this explicit list-mode because we set them independently elsewhere later. (Also, it's possible that this might help mpl optimize for such a state we expect could come later...)

**Comment by agramfort** [comment]
seems to work fine for me !

@Eric89GXL <https://github.com/Eric89GXL> I am just wondering about the
change of behavior on the edges of the head when spatial sampling is poor
(eg with small number of EEG sensors)

**Comment by Eric89GXL** [comment]
I tried to make an example:
```
import mne
evoked = mne.read_evokeds(mne.datasets.testing.data_path() +
                          '/MEG/sample/sample_audvis-ave.fif')[0]
evoked.pick_types(meg=False, eeg=True)
evoked.pick_channels(evoked.ch_names[:10])
evoked.plot_topomap()
```
But even on `master` our (misleading) re-centering / re-scaling makes it weird (would have expected all the electrodes to be at the front of the head, not distributed all over):
screenshot from 2017-04-28 15-55-19
It's not changed much by this PR:
screenshot from 2017-04-28 15-55-06
Let me know if you have a better example...

**Comment by agramfort** [comment]
ok +1 for merge for me then.

**Comment by Eric89GXL** [comment]
@jaeilepp all yours then

**Comment by jaeilepp** [comment]
Thanks @Eric89GXL

**Comment by mmagnuski** [comment]
> But even on master our (misleading) re-centering / re-scaling makes it weird (would have expected all the electrodes to be at the front of the head, not distributed all over)

I also don't like how channels are re-scaled to match the whole head - are there plans to change it?

**Comment by jona-sassenhagen** [comment]
How do you like outlines="skirt"?
